### ISSUE TITLE
Make it work on Buildroot for both py2 and 3

### ISSUE THREAD

**Issue by Evidlo** [crossbario/crossbar#1235]
`process.py` tries to import `shutil`, then falls back to `shutilwhich` if it fails

https://github.com/crossbario/crossbar/blob/master/crossbar/controller/process.py#L38-L43

However, `requirements-min.txt` lists `shutilwhich` as a hard dependency when it shouldn't be.

This causes an error in the `python-crossbar` package for buildroot when building for Python 3.

```
# python
Python 3.6.3 (default, Dec 13 2017, 03:34:46) 
[GCC 6.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
# crossbar start
Traceback (most recent call last):
  File "/usr/bin/crossbar", line 6, in <module>
    from pkg_resources import load_entry_point
  File "usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3037, in <module>
  File "usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3021, in _call_aside
  File "usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3050, in _initialize_master_working_set
  File "usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 655, in _build_master
  File "usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 969, in require
  File "usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 855, in resolve
pkg_resources.DistributionNotFound: The 'shutilwhich>=1.1.0' distribution was not found and is required by crossbar
```

**Comment by meejah** [comment]
How did you install crossbar? I was able to `pip install shutilwhich` on a fresh Python 3.6.3 virtualenv without issue.

**Comment by meejah** [comment]
(Note that `shutilwich` itself will return the built-in implementation on Python's 3.4+)

**Comment by Evidlo** [comment]
Crossbar was built with buildroot.  Right now, buildroot only builds `shutilwhich` [if Python 2 is being used](https://git.busybox.net/buildroot/tree/package/python-shutilwhich/Config.in).

There are two potential solution solutions here:

1.  Buildroot should remove Python 2 as a dependency for `shutilwhich`.
2. Crossbar should [conditionally incude](https://stackoverflow.com/questions/29222269/is-there-a-way-to-have-a-conditional-requirements-txt-file-for-my-python-applica) `shutilwhich` in `requirements-min.txt` only when Python 2 is being used.

**Comment by oberstet** [comment]
`requirements-min.txt` contains the dependencies over all platforms by design, as it is used for input into dependency freezing - which needs to be done again for dependencies over all platform.

we have had conditional dependencies in the past, and it was a pain. and in any case: we should probably strive for a buildroot package that is fully self-contained, not depending on other packages.

we have this with Docker, and snaps. is snapd available for buildroot?

**Comment by yann-morin-1998** [comment]
@meejah said:
> `shutilwich` itself will return the built-in implementation on Python's 3.4+

OK, that's good to know. At least, that leaves us with a way out _as a workaround_.

@meejah @oberstet :
Yet, I find it odd that the code checks _at runtime_ that its dependencies are available,
and even chokes when unneeded dependencies are missing. I would have expected
that the dependency check be done _at configure time_ by `setup.py`, like packages
usually do.

@oberstet :
Buildroot is not a distribution; Buildroot is an embedded buildsystem, not unlike
OpenEmbedded/Yocto or OpenWRT... Buildroot builds packages via cross-compilation
to generate a complete system ready to flash on a device (like a gateway, an STB, or
whatever is called embedded nowadays; graybeard speaking).

There is no "snap" in Buildroot, because it is not a distro. The system that is built is a
complete standalone system all by itself.

(note: not a python expert, I got dragged in by @Evidlo from our IRC channel while
they were trying to add new packages to Buildroot for their use-case.)

**Comment by oberstet** [comment]
@yann-morin-1998 @Evidlo 

pls let me explain the background of this - sorry for a long post for sth that would be trivial to "fix" technically, but I have strong opinions that run deeper here, and I hope you can see why.

> tldr; => rgd the concrete question of shutils and my itches making it a conditional dep, pls see the last paragraph.

---

The overall philosophy is:

Crossbar.io should be treated as a black box server software, with all dependencies hard pinned+hashed and bundled (fully self-contained), and further, we don't want to have an intermediary between us and our users (distro packagers) fuzzing around.

We want to directly ship bit-exact artifacts to our users, and we want end users to allow/enable pulling updates in a deployed system automatically and lights-out.

Docker allows the former. Ubuntu Core allows us to do both. In fact, Ubuntu Core is the only system I am aware of that allows this in a safe way, and including the OS itself.

I would be highly interested in your perspectives here rgd Buildroot (or other "whole system image" things like yocto etc).

How is a Buildroot based system kept up to date? Lights out and automatically in a robust way?

---

Using `requirements-min.txt` is not recommended for production deployments, as the result isn't constant (the exact versions installed of deps will depend on the date at which the installation/build is done - namely, whatever versions of deps happen to be the most recent at said time).

Installation into a global Python environment is also not recommended, as again the result isn't a repeatable artifact.

All of this is by design, because our goal is to ship "bit-exact" packages of Crossbar.io.

Eg the Docker image and the snaps of Crossbar.io X.Y.Z on a given target A can be built in a repeatable way - that is, it doesn't matter _when_ we (or a user) creates the artifacts (Docker image / snap): the result is exactly the same.

The best approximation of the Docker/snap stuff we ship is:

* use a dedicated virtualenv for Crossbar.io
* install with pip from PyPI using [requirements.txt](https://github.com/crossbario/crossbar/blob/master/requirements.txt), NOT `requirements-min.txt`

---

Consequently, we don't ship nor support nor encourage classic distro packages of Crossbar.io, because this leads to non-repeatable deployments (and also non-isolation, because Crossbar.io will be installed into the distros system Python along with potentially a lot other stuff).

In an ideal world, Crossbar.io on Linux as a package would not depend on anything but the Kernel and the C stdlib. Not even OpenSSL or anything.

The Kernel and C stdlib are "the OS", and Crossbar.io would only rely on that and not anything else that might come on the target.

Docker and snaps allow us to approximate above pretty closely. Classic Linux distros are orthogonal to our goals.

Regarding full system image Linux (like buildroot, yocto, ..): the way to go (mid/long-term) IMO is to create a fully self contained package for Crossbar.io, that bundles its own Python and everything.

That is, baking a buildroot image with Crossbar.io should be possible with selecting a "Crossbar.io package", and nothing else, not even Python (as that should be bundled in said package).

---

On the topic of conditional deps, there are two aspects here:

1. we don't want to have "optional features" in Crossbar.io, because that splitters up what users have in their hands (we did that in the past, and it wasn't pretty nor did it fit the "100% bit exact repeatability" goal).  

For example, Crossbar.io now ships with Tor support. Always.

Distro packagers have started to [patch away](https://gitweb.gentoo.org/repo/gentoo.git/diff/net-misc/crossbar/files/crossbar-17.9.1-Revert-Torservices.patch?id2=10f2b985a5ec30cf1a2b299dc0a9658996fa1794&context=9&ignorews=1) this (and other features): I absolutely disagree, and will oppose everywhere I can, though of course we cannot stop those guys doing what they do.

2. polyfills (eg shutils)

This is somewhat different, as it doesn't splitter Crossbar.io feature-wise.

However, here we are running into a limitation of Python packaging: **it is not possible (to my knowledge) to combine pinned+hashed deps and conditional deps**.

If that would be possible, I'd be +1 on doing it.

**Comment by yegorich** [comment]
@Evidlo shutilwhich is now installable for both Python versions, so the main issue is the handling of ipaddress and enum34 packages.

**Comment by yegorich** [comment]
@oberstet Tor support will be somewhat problematic on such distros as BR, OpenWrt, Yocto etc. because it introduces `hmm` dependency, that requires `scipy`. And here the fun begins https://github.com/scipy/scipy/issues/8226 :-(

**Comment by yegorich** [comment]
@oberstet forget about my last comment. hmm is not a dependency, so no need to worry about Tor support. Sorry for the noise.

**Comment by oberstet** [comment]
dup https://github.com/crossbario/crossbar/issues/1146

**Comment by Evidlo** [comment]
[Here's](https://gist.github.com/Evidlo/998c2bccca503e3947b430afc7a0e4ed) a diff of the changes I had to make to various buildroot packages to get crossbar running on a Pi 3 (as of December 2017).

According to @yegorich, line 79 of my diff is already in buildroot.

**Comment by yegorich** [comment]
@Evidlo have you tried the latest BR (master branch)? I could successfully invoke `crossbar version` on both Python2 and Python3 images.

Autobahn doesn't need Twisted as it is an "extra" dependency.

As for 2to3 option you should submit a patch to BR's mailing list. Though autobahn has no problems being built and installed without this option.

I'll have to check glibc dependency.

**Comment by yegorich** [comment]
As for uClibc I had so far only one issue with watchdog module as it tries to import libc and it cannot find it. Because it searches either for `libc.so` or `libc.so.6` and uClibc has following symlinks `libc.so.0` and `libc.so.1`. So watchdog should be fixed, i.e. this issue https://github.com/gorakhargosh/watchdog/issues/253 should be reopened.

**Comment by yegorich** [comment]
@oberstet I think for now there is nothing for you to do. When I bump crossbar I just provide a [patch](https://git.busybox.net/buildroot/tree/package/python-crossbar/0002-Rework-runtime-dependencies.patch) for requirements file.

@Evidlo please provide your BR configuration (deconfig) where you experience ValueErrors. Please update to the latest master first.

Meanwhile I'll try to prepare a crossbar test for BR's gitlab testing facility.

**Comment by oberstet** [comment]
looking at @yegorich 's patch, I noticed:

```
-h2<3.0.0
+h2>=3.0.1
```

I forgot why, but I'm pretty sure we had reasons for that, and that reason might be gone with later versions of h2 .. so we should revisit.

@meejah do you remember?

**Comment by Evidlo** [comment]
@yegorich Try starting crossbar with a configuration.  I don't have the hardware to test on right now, but I remember there was a point where everything seemed to work until I actually launched crossbar with an autobahn application.

**Comment by oberstet** [comment]
Can you guys point me to some "common" hardware board that is supposed to run Crossbar.io?

We have stuff like the Pi 3 or Samsung Artik 710 (both quite beefy and using Ubuntu Core), but I don't have a lot clue about what people use with Buildroot ..

We have a hardware partner (Blacbird) around the corner here that is familiar with Buildroot - using it with AutobahnPython on http://www.opossom.com/english/products-processor_boards-opos6ul.html

If we purchase a board, I could forward to them for trying it out with not only AB, but the full CB ..

[sidenote: anyone on embedded world 2018? we could meet .. we'll be exhibiting]

**Comment by meejah** [comment]
Looks like the "< 3.0.0" came from https://github.com/crossbario/crossbar/pull/1052 -- says "because that was too new for Twisted". I would assume that's all been worked out now as Twisted depends on any h2 version 3 now (https://github.com/twisted/twisted/commit/3366f2010b33c28b33f7cd80dabcbe6988f350b9)

**Comment by oberstet** [comment]
@meejah thanks for digging! still puzzled: h2 3.0.1 is (still) the latest h2 release, and there isn't even a 4 branch: https://github.com/python-hyper/hyper-h2

since @Lukasa is original author of h2, and also Twisted committer (also on the particular tx change you link) .. maybe he can shed some light? why < 4?

anyway, pragmatically, replicating what tx does sounds good to me: `h2 >= 3.0, < 4.0`

and that is still the current version spec on tx trunk:

https://github.com/twisted/twisted/blob/trunk/src/twisted/python/_setup.py#L110

**Comment by meejah** [comment]
I'm sure it's just hedging against any possible breaking changes in 4.x (but yeah, we should just do whatever is in Twisted since that the only reason for the dependency).

**Comment by Lukasa** [comment]
Yup, that’s just hedging against further breaking changes.

**Comment by oberstet** [comment]
@Lukasa hey Cory, long time no talk;) hope you are all fine. thx for confirmation!

**Comment by oberstet** [comment]
ok, I removed the following dependencies on master:

* enum34
* shutilwhich
* ipaddress

and set `h2 >= 3.0, < 4.0`.

this should avoid the hassles on py 3 builds.

> these changes will be on crossbar 18.5.1, which will officially drop python 2 support. in return, we plan to use the new py 3.5+ type hints to increase robustness of the code base
